src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -145,75 +145,33 @@ public void Shutdown()
-                {
-                    // Check if cancellation was requested while this was waiting in the queue
-                    if (cancellationToken.IsCancellationRequested)
-                    {
-                        completion.SetCanceled();
-
-                        return;
-                    }
-
-                    // If we weren't able to get a corresponding context for this request (for example, we
-                    // couldn't map a doc request to a particular Document, or we couldn't find an appropriate
-                    // Workspace for a global operation), then just immediately complete the request with a
-                    // 'null' response.  Note: the lsp spec was checked to ensure that 'null' is valid for all
-                    // the requests this could happen for.  However, this assumption may not hold in the future.
-                    // If that turns out to be the case, we could defer to the individual handler to decide
-                    // what to do.
-                    if (context == null)
-                    {
-                        completion.SetResult(default);
-                        return;
-                    }
-
-                    try
-                    {
-                        var result = await handler.HandleRequestAsync(request, context.Value, cancellationToken).ConfigureAwait(false);
-                        completion.SetResult(result);
-                    }
-                    catch (OperationCanceledException ex)
-                    {
-                        completion.TrySetCanceled(ex.CancellationToken);
-                    }
-                    catch (Exception exception)
-                    {
-                        // Pass the exception to the task completion source, so the caller of the ExecuteAsync method can react
-                        completion.SetException(exception);
-
-                        // Also allow the exception to flow back to the request queue to handle as appropriate
-                        throw new InvalidOperationException($"Error handling '{methodName}' request: {exception.Message}", exception);
-                    }
-                }, requestCancellationToken);
+                requestCancellationToken);
 
             var didEnqueue = _queue.TryEnqueue(item);
 
             // If the queue has been shut down the enqueue will fail, so we just fault the task immediately.
             // The queue itself is threadsafe (_queue.TryEnqueue and _queue.Complete use the same lock).
             if (!didEnqueue)
             {
-                completion.SetException(new InvalidOperationException($"{_serverName} was requested to shut down."));
+                item.CompletionSource.SetException(new InvalidOperationException($"{_serverName} was requested to shut down."));
             }
 
-            return completion.Task;
+            return item.CompletionSource.Task;
```

Comment:
❓ Does this execute inside Visual Studio? Is it expected to cooperate with JTF?

Reply:
it does execute in VS, however I believe it shouldn't need jtf since
1.  JTF does not keep context across the streamjsonrpc boundary, so even if the client is blocking the UI, using JTF here won't help
2.  LSP handlers should not be using the UI thread.  If they do it is a bug.  There currently one that does, but because of 1) it is broken in any scenario where the LSP server is running on the same machine (https://github.com/dotnet/roslyn/issues/51951)